Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.ts

```diff
@@ -0,0 +1,355 @@
+
+type BocAutoCompleteReferenceValue_UpdateResult = { Value: Nullable<BocAutoCompleteReferenceValue_Item> }
+
+class BocAutoCompleteReferenceValue //TODO RM-7715 - Make the TypeScript classes BocReferenceValue and BocAutoCompleteReferenceValue inherit from BocReferenceValueBase
+{
+  private _itemBackUp: Nullable<BocAutoCompleteReferenceValue_Item>;
+  private _isInvalidated: boolean;
+  private _command: Nullable<JQuery>;
+  private _commandBackUp: Nullable<JQuery>;
+  private _selectListID: string;
+  private _informationPopUpID: string;
+
+  constructor (
+    baseID: Nullable<string>,
+    combobox: JQuery, textbox: JQuery, hiddenField: JQuery, button: JQuery, command: Nullable<JQuery>, controlServiceUrl: string,
+    completionSetCount: number, dropDownDisplayDelay: number, dropDownRefreshDelay: number, selectionUpdateDelay: number,
+    searchStringValidationInfo: BocReferenceValueBase_SearchStringValidationInfo,
+    nullValueString: string,
+    isAutoPostBackEnabled: boolean,
+    searchContext: object,
+    isIconUpdateEnabled: boolean,
+    iconContext: BocReferenceValueBase_IconContext,
+    commandInfo: Nullable<BocReferenceValueBase_CommandInfo>,
+    resources: BocReferenceValueBase_Resources)
+  {
+    ArgumentUtility.CheckTypeIsString('baseID', baseID);
+    ArgumentUtility.CheckNotNullAndTypeIsObject('combobox', combobox);
+    ArgumentUtility.CheckNotNullAndTypeIsObject('textbox', textbox);
+    ArgumentUtility.CheckNotNullAndTypeIsObject('hiddenField', hiddenField);
+    ArgumentUtility.CheckNotNullAndTypeIsObject('button', button);
+    ArgumentUtility.CheckTypeIsObject('command', command);
+    ArgumentUtility.CheckNotNullAndTypeIsString('controlServiceUrl', controlServiceUrl);
+    ArgumentUtility.CheckNotNullAndTypeIsNumber('completionSetCount', completionSetCount);
+    ArgumentUtility.CheckNotNullAndTypeIsNumber('dropDownDisplayDelay', dropDownDisplayDelay);
+    ArgumentUtility.CheckNotNullAndTypeIsNumber('dropDownRefreshDelay', dropDownRefreshDelay);
+    ArgumentUtility.CheckNotNullAndTypeIsNumber('selectionUpdateDelay', selectionUpdateDelay);
+    ArgumentUtility.CheckNotNullAndTypeIsObject('searchStringValidationInfo', searchStringValidationInfo);
+    ArgumentUtility.CheckNotNullAndTypeIsString('nullValueString', nullValueString);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('isAutoPostBackEnabled', isAutoPostBackEnabled);
+    ArgumentUtility.CheckNotNullAndTypeIsObject('searchContext', searchContext);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('isIconUpdateEnabled', isIconUpdateEnabled);
+    if (isIconUpdateEnabled)
+      ArgumentUtility.CheckNotNullAndTypeIsObject('iconContext', iconContext);
+    ArgumentUtility.CheckTypeIsObject('commandInfo', commandInfo);
+    ArgumentUtility.CheckNotNullAndTypeIsObject('resources', resources);
+
+    const BackupItemData = (uniqueIdentifier: string, displayName: string) =>
```

Comment:
I don't have the sufficient experience with JavaScript/TypeScript to engage in such a conversation, please ask @a-ctor, who told me to do it like this in the first place

Reply:
As a reference, this is what he wrote as a comment, maybe that's even enough:

> We should not modify the existing functions but instead use the this parameter. Since those functions also use the parameters of the function it gets a bit tricky since we cannot convert them to methods.

> Instead we can convert the methods into variables and assign lambdas to them. e.g. const RestoreCommand = () => { ... }. This allows them to use this instead of introducing a new parameter. Since those methods are used above we need to moved them above their usage.

Reply:
@Stefan-Ilic I understand where @a-ctor is coming from (or I think I do) but I don't think is is necessary. Most usages already access fields based on `this` so calling functions based on `this` should be the same. What remains are one or two usage sites where we have nested closures. there, we could just capture `this` in a variable and call the function based on this variable.

This design allows us to keep using instance functions, does not require moving code blocks around and is generally more consistent with the rest of the migration.

What I did notice (but I don't think we should change this right now) that the anonynous functions added to the jquery object (see `updateResult`, `invalidateResult`, ...) could perhaps also be changed to instance functions which are then simply added as a delegate to the jquery object. this would perhaps result in an even cleaner design since it creates better defined chunks of code instead of this jquery-spagetti we have right now.

Reply:
I will make the changes you requested, but I do hope you understand how frustrating it is to repeatedly work in and revert conflicting change requests from two different reviewers

Reply:
@a-ctor, @Stefan-Ilic is correct about the missing alignment. This type and it's constraints is simply unfortunate. In the end we want the most amount of statically typed code and try to avoid closures when we can write clean code in a differnt way.

Reply:
There are two problems in this file: 
 a) parameters are used in the helper functions and 
 b) due to the class syntax local variables cannot be accessed in the helper functions without capturing.

Here are some possible solutions:

```js
const ClearError = () =>
{
  if (textbox.hasClass('error'))
  {
    textbox.removeClass ('error');
    textbox.trigger ('hideInformationPopUp');
  }
}

// Usage:
ClearError(); // must be after the declaration
```

Creating closures solves both problems, but has the downside of needing to move the declarations before its usage. This was the current solution I came up with.

```js
// this.textbox has to be set in the constructure

// Definition
private ClearError()
{
  if (this.textbox.hasClass('error'))
  {
    this.textbox.removeClass ('error');
    this.textbox.trigger ('hideInformationPopUp');
  }
}

// Usage:
this.ClearError();
// but if it needs to be passed as a delegate
passDel(this.ClearError.bind(this))
```

Another solution would be to convert the helper methods into instance methods and set any parameters that are used in those methods as fields. But whenever the method is passed as a delegate the `this` parameter must be fixed in the method using `.bind(this)`. Otherwise `this` is not set when the method is called.

```js
// this.textbox has to be set in the constructure

// Definition
ClearError = () =>
{
  if (this.textbox.hasClass('error'))
  {
    this.textbox.removeClass ('error');
    this.textbox.trigger ('hideInformationPopUp');
  }
};

// Usage:
this.ClearError();
```

Probably the best solution would be to do the previous solution but declare the helper functions as closures. This has the benefit that the function already captures `this` so you can just use it without any `.apply`. Parameters still need to be set to instance fields, but you also no do not have to reorder members.

Reply:
After rereading, yes, we have a couple (~4) captured parameters that would have to be converted to fields, but the closures I was originally referring to, those don't get used via delegates and should be easily convertible to instance functions. 
* BackupItemData
* UpdateCommand
* GetDropDownSearchStringForValidInput 
* (I hope I got all)

Thanks for the reminder regarding the delagates, that would just be syntax clenaup anway and create more noise than it would fix right now. Not saying that we won't do it later.